File: src/main/java/com/mock_json/mock_api/models/Url.java
package com.mock_json.mock_api.models;

import jakarta.persistence.*;
import lombok.*;

import org.hibernate.annotations.SoftDelete;
import org.hibernate.annotations.UpdateTimestamp;

import com.fasterxml.jackson.annotation.JsonBackReference;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonManagedReference;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.mock_json.mock_api.enums.StatusCode;

import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.NaturalId;

import java.time.LocalDateTime;
import java.util.List;

@Entity
@Table(name = "urls")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
@SoftDelete
public class Url {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = true,columnDefinition = "TEXT")
    private String description;

    @Column(nullable = false)
    private String name;

    // for mocking rate limits
    @Column(nullable = true)
    private  Integer requests;

    @Column(nullable = true)
    private  Long time;

    @Column(nullable = false, unique = false)
    @NaturalId
    private String url;

    @Column(nullable = false)
    @Enumerated(EnumType.STRING) 
    private StatusCode status; 

    @CreationTimestamp
    @Column(nullable = false, updatable = false)
    @JsonIgnore
    private LocalDateTime createdAt;

    @UpdateTimestamp
    @Column(nullable = false)
    @JsonIgnore
    private LocalDateTime updatedAt;

    @ManyToOne(fetch = FetchType.EAGER)
    @JoinColumn(name = "project_id")
    @JsonBackReference
    @NaturalId
    private Project project;

    @OneToMany(mappedBy = "urlId", cascade = CascadeType.ALL, fetch = FetchType.EAGER)
    @JsonProperty("mock_content_list")
    @JsonManagedReference
    private List<MockContent> mockContentList;

}

File: internal/models/url.go
package models

import (
	"time"

	"gorm.io/gorm"
)

// Url represents a URL in the system.
type Url struct {
	ID              uint           `gorm:"primarykey" json:"id"`
	CreatedAt       time.Time      `json:"created_at"`
	UpdatedAt       time.Time      `json:"updated_at"`
	DeletedAt       gorm.DeletedAt `gorm:"index" json:"-"`
	Description     string         `json:"description"`
	Name            string         `gorm:"uniqueIndex:idx_url_name_project_id;not null" json:"name"`
	URL             string         `gorm:"column:url;uniqueIndex:idx_url_url_project_id;not null" json:"url"`
	ProjectID       uint           `gorm:"uniqueIndex:idx_url_name_project_id;uniqueIndex:idx_url_url_project_id;not null" json:"project_id"`
	Project         *Project       `gorm:"foreignKey:ProjectID" json:"project,omitempty"`
	MockContentList []MockContent  `gorm:"foreignKey:URLID" json:"mock_content_list,omitempty"`
}

// TableName specifies the table name for the Url model.
func (Url) TableName() string {
	return "urls"
}
